### PR TITLE
updated tensorflow 2.0; protobuf 3.8.0

### PR DESCRIPTION
**Pull request by smuzaffar**
It is based on the work of @riga
- Update Tensorflow to version 2.0.0
  - build with `std=c++17`
  - Combine the TF build recipe in one file. So that both TF Py2 and TF Py3 are build using same way
  - Run bazel in batch mode to avoid starting bazel server
  - Added/included extra python package dependencies
  - For TF PY3 , only build pip package. No need to build tensorflow cc, c interface
- Protobuf 3.8.0 (used by tensorflow 3.8.0)
- Bazel version 0.26.1
- Onnxruntime to use Protobuf 3.8.0
  - Protobuf 3.8.0+: Commment out LimitingInputStream ( https://github.com/cms-externals/onnxruntime/commit/0f048da7774428d5fb2c4c808fc5820809ab24b8 )

We need cmssw updates to use this version. Specially  regenerating DQMServices/Core/src/ROOTFilePB.proto for new protobuf and update PhysicsTools/Tensorflow

### CODE REVIEW

**Comment by smuzaffar** [comment]
the python tests we have works with TF 2.0
```
pbash-4.2$ python ./test_deepjet.py 
CMSSW: CMSSW_11_1_X_2019-12-12-1100
TF   : 2.0.0

loading graph from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/cms/cmssw-patch/CMSSW_10_6_0_patch1/external/slc7_amd64_gcc700/data/RecoBTag/Combined/data/DeepFlavourV03_10X_training/constant_graph.pb
done

starting session
2019-12-16 17:59:38.430693: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX
done

evaluating the session
done

DeepJet outputs:
[array([[0.19014773, 0.00817215, 0.05376633, 0.17345323, 0.14495747,
        0.42950305]], dtype=float32)]
```

**Comment by smuzaffar** [comment]
please test
just to make sure that externals build successfully

**Comment by riga** [comment]
@smuzaffar I am currently building this PR on a machine with GPUs attached and will open a PR to CMSSW once the changes to PhysicsTools/Tensorflow are complete.

File: tensorflow-sources.file [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+# copy headers from downloaded dependencies
+copy_headers() {
+  for header_file in $(find $1/$2 -name *.h | sed "s|$1/||") ; do
+    header_dir="${incdir}/$(dirname ${header_file})"
+    mkdir -p "${header_dir}"
+    cp -p "${header_file}" "${header_dir}/"
+  done
+}
+copy_headers "$PWD" tensorflow/compiler
```
Comment (riga) [inline tensorflow-sources.file:149]: Hi @smuzaffar ,
your changes compile on my GPU machine, but I noticed that some headers are missing when I compile the updated sessions (e.g. `tensorflow/core/profiler/lib/profiler_session.h`). It seems to work though if we add

```
copy_headers "$PWD" tensorflow/core/profiler/internal
copy_headers "$PWD" tensorflow/core/profiler/lib
```
below this line.

Comment (smuzaffar) [reply tensorflow-sources.file:149]: thanks @riga, I have updated the PR

**Comment by smuzaffar** [comment]
@riga , were you able to run cmssw tests?

**Comment by riga** [comment]
@smuzaffar The tests are linked to libtensorflow_cc.so.2 which is currently missing. I created PR #5435 which adds them so that (most) tests work.

<hr />

Except for the `tfadd_t.cpp` test, all of them compile but do not succeed yet due to some trivial change from TF v1 to v2. This should be fairly easy to fix. I'm on it.

<hr />

`tfadd_t.cpp` seems a bit more tricky. I get this while compiling:

```
2019-12-17 15:20:08.640612: F tensorflow/compiler/xla/service/cpu/cpu_compiler.cc:912] Check failed: entry_function->getName() == entry_point_name
/bin/sh: line 1: 12674 Aborted                 /data/hgcal-0/user/mrieger/tfcmsdist/slc7_amd64_gcc820/external/tensorflow/2.0.0-cms/bin/tfcompile --target_features="+sse4.2" --graph=/afs/cern.ch/user/m/mrieger/cmssw/tf20/CMSSW_11_0_X_2019-12-16-1100/src/PhysicsTools/TensorFlow/test/test_graph_tfadd.pb --config=/afs/cern.ch/user/m/mrieger/cmssw/tf20/CMSSW_11_0_X_2019-12-16-1100/src/PhysicsTools/TensorFlow/test/test_graph_tfadd.config.pbtxt --entry_point=__tensorflow_test_graph_tfadd --cpp_class=test_graph_tfadd --target_triple=x86_64-unknown-linux-gnu --out_header=tmp/slc7_amd64_gcc820/include/tfadd_t/src/PhysicsTools/TensorFlow/test/test_graph_tfadd/header.h --out_metadata_object=tmp/slc7_amd64_gcc820/src/PhysicsTools/TensorFlow/test/tfadd_t/test_graph_tfadd/tfcompile_metadata.o --out_function_object=tmp/slc7_amd64_gcc820/src/PhysicsTools/TensorFlow/test/tfadd_t/test_graph_tfadd/tfcompile_function.o
```

This might be caused by the old `test_graph_tfadd.pb` file that is used in the test. Do you remember how this exactly was updated with `tfcompile` / `protoc`?

**Comment by smuzaffar** [comment]
@riga , do you have cmssw changes which we can test? `tfadd_t.cpp` test can be disabled if not working. We might have to regenerate the .pb file for newer version

**Comment by riga** [comment]
@smuzaffar Yes, I just finished updating the sessions and adjusting the tests. Except for `tfadd_t.cpp` they all pass. I pushed all changes to [riga:tensorflow2](https://github.com/riga/cmssw/tree/tensorflow2) (based on `CMSSW_11_1_X_2019-12-19-1100`) in case you want to check them out.

I will now try to regenerate DQMServices/Core/src/ROOTFilePB.proto (edit: done, this was easier than expected), add another commit, rebase, and eventually open a cmssw PR if you don't see any crucial issues.

Btw, I also started to adjust the interface to simplify device placement (CPU vs. GPU), however, this is not yet included in the branch above. I guess we target GPU support in another cmssw/dist PR?

**Comment by riga** [comment]
@smuzaffar I investigated why the `tfadd_t.cpp` test fails. This is of course unrelated to this PR, but is there a reason for the `__tensorflow_` prefix of the`--entry_point` argument in the [cmssw-config/SCRAM/GMake/Makefile.dnn](https://github.com/cms-sw/cmssw-config/blob/4d6cd0fef1e12ae9f3e4a4cd5fda0917a940e085/SCRAM/GMake/Makefile.dnn#L42) build rule?

```
$(TFCOMPILE) --target_features="$(DNN_TARGET_FEATURE)" --graph=$$($(1)_DNN_PB) \
  --config=$$($(1)_DNN_CONFIG_PBTXT) --entry_point=__tensorflow_$(3) ... etc
```

I pasted the error message [above](#issuecomment-566580499).

The `tfcompile` command works without the prefix.

**Edit**: It looks like the entry point is no longer allowed to start with two underscores. `--entry_point=_tensorflow_$(3)` works just fine.

**Comment by smuzaffar** [comment]
@riga thanks for looking in to it. I have no idea  about tfcompile. When we integrated this rules in scram then I was told to run tfcompile with `--entry_point=__tensorflow_${DNN_NAME}` flag. May be it is not needed any more with TF v2. If you can make a cmssw PR with your changes then I can also provide a cmssw-config (scram build rules) PR to fix this.

**Comment by mrodozov** [comment]
@riga search for this in the mail (by subject)
Stuck tests for PR 25889
you were in cc, there's the recipe how to remake the .pb files.

**Comment by smuzaffar** [comment]
closing in favor of https://github.com/cms-sw/cmsdist/pull/5509

**Pull request closed without merging**